pandas/core/frame.py

```diff
@@ -416,44 +416,29 @@ def _init_dict(self, data, index, columns, dtype=None):
         Needs to handle a lot of exceptional cases.
         """
         if columns is not None:
-            columns = _ensure_index(columns)
+            arrays = Series(data, index=columns, dtype=object)
+            data_names = arrays.index
```

Comment:
this will be a perf issue

Reply:
Maybe... but right now it seems to be worse...

```      before           after         ratio
     [d163de70]       [f7447b3f]
-      47.9±0.3ms       43.5±0.4ms     0.91  frame_ctor.FromDicts.time_frame_ctor_nested_dict
-      31.0±0.1ms       28.1±0.3ms     0.91  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('BusinessDay', 2)
-        31.3±1ms       28.2±0.2ms     0.90  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('BDay', 2)
-      31.8±0.3ms       28.0±0.4ms     0.88  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('CustomBusinessDay', 2)
-      32.8±0.3ms       28.2±0.2ms     0.86  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('Day', 1)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.

```

Reply:
There does seem to be a performance loss on very small dfs. E.g. for ``pd.DataFrame(data)`` with ``data = {1 : [2], 3 : [4], 5 : [6]}`` I get results around 530 µs for  per loop before and 570 µs after. So we are talking about a ~10% gain on large dfs vs. a ~7.5% loss on small dfs.

... or I can avoid that ``Series`` and sort manually, at the cost of a bit of added complexity, probably ~10 LoCs.

Reply:
uhm... those asv results also seem pretty unstable:
```
      before           after         ratio
     [d163de70]       [f7447b3f]
+      30.7±0.2ms         41.1±3ms     1.34  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('QuarterBegin', 1)
+      29.4±0.1ms         39.1±4ms     1.33  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('CBMonthBegin', 2)
+      30.7±0.7ms         40.4±3ms     1.32  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('BDay', 2)
+      31.1±0.7ms         39.1±5ms     1.26  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('SemiMonthEnd', 2)
+      30.4±0.1ms         38.0±4ms     1.25  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('Hour', 2)
-        33.8±1ms       30.4±0.8ms     0.90  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('Micro', 1)
-      48.3±0.6ms       42.6±0.8ms     0.88  frame_ctor.FromDicts.time_frame_ctor_nested_dict
-        23.8±1ms       20.5±0.2ms     0.86  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('FY5253Quarter_1', 2)
-      41.2±0.7ms         30.4±2ms     0.74  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('CustomBusinessHour', 2)
-      8.35±0.9ms      6.05±0.01ms     0.72  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('FY5253_2', 2)
-        42.4±2ms       30.5±0.5ms     0.72  frame_ctor.FromDictwithTimestampOffsets.time_frame_ctor('BMonthEnd', 2)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

I'll try to sort manually and see how it goes.

Reply:
i actually doubt we have good benchmarks on this you are measuring the same benchmark here

we need benchmarks that contruct with different dtypes

and reducing code complexity is paramount here (though of course don’t want to sacrifice perf)